pkg/model/time_series_builder.go

```diff
@@ -25,59 +26,126 @@ func NewTimeSeriesBuilder(by ...string) *TimeSeriesBuilder {
 }
 
 func (s *TimeSeriesBuilder) Init(by ...string) {
-	s.labelsByFingerprint = map[model.Fingerprint]string{}
 	s.series = make(seriesByLabels)
 	s.labelBuf = make([]byte, 0, 1024)
 	s.by = by
+	s.exemplarBuilders = make(map[string]*ExemplarBuilder)
 }
 
-func (s *TimeSeriesBuilder) Add(fp model.Fingerprint, lbs Labels, ts int64, value float64, annotations schemav1.Annotations) {
-	labelsByString, ok := s.labelsByFingerprint[fp]
+// Add adds a data point with full labels.
+// The series is grouped by the 'by' labels, but exemplars retain full labels.
+func (s *TimeSeriesBuilder) Add(fp model.Fingerprint, lbs Labels, ts int64, value float64, annotations schemav1.Annotations, profileID string) {
+	s.labelBuf = lbs.BytesWithLabels(s.labelBuf, s.by...)
+	seriesKey := string(s.labelBuf)
+
 	pAnnotations := make([]*typesv1.ProfileAnnotation, 0, len(annotations.Keys))
 	for i := range len(annotations.Keys) {
 		pAnnotations = append(pAnnotations, &typesv1.ProfileAnnotation{
 			Key:   annotations.Keys[i],
 			Value: annotations.Values[i],
 		})
 	}
-	if !ok {
-		s.labelBuf = lbs.BytesWithLabels(s.labelBuf, s.by...)
-		labelsByString = string(s.labelBuf)
-		s.labelsByFingerprint[fp] = labelsByString
-		if _, ok := s.series[labelsByString]; !ok {
-			s.series[labelsByString] = &typesv1.Series{
-				Labels: lbs.WithLabels(s.by...),
-				Points: []*typesv1.Point{
-					{
-						Timestamp:   ts,
-						Value:       value,
-						Annotations: pAnnotations,
-					},
+
+	series, exists := s.series[seriesKey]
+	if !exists {
+		s.series[seriesKey] = &typesv1.Series{
+			Labels: lbs.WithLabels(s.by...),
+			Points: []*typesv1.Point{
+				{
+					Timestamp:   ts,
+					Value:       value,
+					Annotations: pAnnotations,
 				},
-			}
-			return
+			},
 		}
+	} else {
```

Comment:
Nit: I would find that a lot more readable, when we would get rid of the `else` and always append the point. (and obviously not append the point in `if !exists`.